duktape.js

```diff
@@ -26,6 +26,9 @@ var dukKey = (function () {
     });
     var dukVersion = Number(stdout);
     console.log('Duktape version is: ' + dukVersion);
+    if ((dukVersion % 100) == 99) {
+        dukVersion++;  // Treat e.g. 2.2.99 (built from master) as 2.3.0 for testing
```

Comment:
Can you elaborate on why this is in here, or why `2.2.99` would be anything more special than the patch version after `2.2.98`?

Reply:
Right after 2.2.0 has been released, Duktape master switches to 2.2.99 which is used until the 2.3.0 release (http://duktape.org/guide.html#versioning.4). While no version is correct for a master build, the .99 patch is used because the build is closer to 2.3.0 than any 2.2.x maintenance version for comparison purposes.


When one runs compat-table/duktape.js with the 2.2.99 release, prior to this change it will appear as Duktape 2.2 and compare against `duktape2_2` results. This is a bit useless because there's no good way to fix the data entries; one would need to alter 2.2 results.

With this change it will compare against `duktape2_3` results and allow one (in practice me) to maintain the to-be-released 2.3.0 results incrementally.

This is not an important change, but before this, I had to manually force dukVersion in duktape.js for testing and maintaining the results incrementally.

Reply:
In that case, duktape is not following semver; it should be using prerelease builds for that. `.100` is no more special than `.1` or `.1000`, and misusing `99` like that means semver-minor changes appear in the 2.2 line after .0.

We'll have to alter the entire way we handle duktape versions inside compat-table unless it changes this policy.

Reply:
Note that builds from master are not releases at all, so Duktape *releases* do follow semver but master builds are not releases.

Reply:
A bit more background, DUK_VERSION is a number of the form 10203 (for 1.2.3) and between releases I guess there are roughly the following options (assuming master build after 2.2.0 but before 2.3.0):
- 2.2.99: current policy, patch level never reaches .99. It's possible to detect an unofficial build from the version number, and 'version >= 2.3.0' check won't catch the unofficial build which is good.
- 2.3.0: this would also be workable, but official release would have to be 2.3.1 to avoid confusion. But this would also break semver: 2.3.0 -> 2.3.1 might entail arbitrary changes (not just patch level changes), e.g. if the 2.3.0 was built right after 2.2.0.
- Alternate releases, e.g. 2.2.0 official release, 2.3.0 reserved for development, 2.4.0 official release, etc. This would be closest to fulfilling semver, but even with this 2.3.x patch level guarantees would be broken because development between minor versions often involves multiple individual changes deserving a minor version bump.

For a numeric version like 1.2.3, with no space for `-alpha` or similar qualifiers, it's not really possible to fulfil semver as far as I can see it, unless one bumps the version literally for every commit going into (unreleased) master, which is very inconvenient. So, that's why the current compromise is to use strict semver for releases only, not manual builds from Duktape master.

Reply:
No, it wouldn't - you'd use "prereleases", which are part of the semver spec and defined for exactly this purpose. You'd do `2.3.0-alpha.0` or `2.3.0-rc.1` etc, and then `2.3.0` when ready.

Using `.99` makes no sense whatsoever.

Reply:
Given that DUK_VERSION is a number of the form major*10000 + minor*100 + patch, there's no way to add a suffix like that. The reason it is a number is to allow C programmers to do things like:

```c
#if (DUK_VERSION >= 20300)
/* Use Duktape 2.3 features */
#else
/* Older */
#endif
```

C preprocessor cannot do version comparisons from strings, which is why version number defines are common for C libraries.

Reply:
Since that's the case, that means duktape does not follow semver, and the versioning logic and automated scripts for it in this repo were allowed based primarily on following semver.

Reply:
Does semver require explicit versioning for every *unreleased* commit in repository?

Because that's what we're talking about here: there are no releases on duktape.org with a .99 patch level, and there won't be any releases like that.

Reply:
Fair point; in that case, let's find a better hack than a ".99" version; or at least let's not let it pollute this repo.

Reply:
Like I explained, that "hack" is the best compromise in practice. Something must go into DUK_VERSION, and 20299 is better than 20300, at least based on user feedback.

But I can remove that check if it makes things better. It just means running duktape.js for a development build without manual tweaks gives nonsensical results. I fail to see why that's better.

Reply:
Personally I find that reserving the `.0` version for this, and only ever releasing `.1` and higher, would be a much cleaner patch than breaking semver just to accomodate deficiencies in C.

Reply:
But that would also break semver: the changes from 2.3.0 -> 2.3.1 would in many cases be more significant than patch level changes?

Reply:
hm, also a fair point. not really sure what to do here.

Reply:
One option is to detect the .99 patch level and refuse to run, indicating the cause: you're running a development version which exists outside the semver version number space and it's not obvious which results you'd want to compare against.

duktape.js could then provide an option to force the version to be used for results purposes. I could then run 2.2.99 as `nodejs duktape.js --force-duk-version 20300` to update the 2.3 results incrementally. This would avoid the need to manually hack the source file.

But I'm fine with whatever really -- I can keep on editing duktape.js manually every time I run the tests, and I doubt many others will run a Duktape development build against the test set.